### PR TITLE
[WIP][ENH] Scatterplot, HeatMap, TreeGraph, ConfusionMatrix and Unsupervised widgets: Output Flagged Data 

### PR DESCRIPTION
**Pull request by VesnaT**

### CODE REVIEW

**Comment by janezd** [comment]
I haven't yet checked the content of the PR (obviously), but before we merge it we should perhaps talk about the signal name again. I don't think that the verb "to flag" appears in the dictionary with that meaning. Even the use of word "flag" for some kind of marker is IMHO limited to computer science, so the name "flagged data" wouldn't mean anything to an outsider.

What is wrong with "marked"?

Vesna, sorry if this will require some (hopefully trivial) refactoring.

**Comment by VesnaT** [comment]
No worries.. I've intentionally made this PS 'small' (only four widgets are modified).

File: Orange/misc/flagged_data.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    and is generated using name parameter, appended by smallest possible index.
+
+    :param names: list
+    :param name: str
+    :return: str
+    """
+    if name not in names:
+        return name
+    counts = [int(re.match(r"(" + name + " )(\d{1,}$)", n).group(2))
+              for n in names if re.match(r"(" + name + " )(\d{1,}$)", n)] + [1]
```
Comment (janezd) [inline Orange/misc/flagged_data.py]: Something like this:
```
counts = max((int(mo.group(2))
              for mo in re.finditer(r"(" + name + " )(\d{1,}$)", n)),
             default=0)
```

Comment (janezd) [inline Orange/misc/flagged_data.py]: Matter of taste: I'd write `r"({})(\d{{1,}}$".format(name)`

File: Orange/tree.py [hunk 2]
```diff
@@ -228,6 +228,11 @@ def get_instances(self, nodes):
         if subsets:
             return self.instances[np.unique(np.hstack(subsets))]
 
+    def get_indices(self, nodes):
```
Comment (janezd) [inline Orange/tree.py]: `get_instances` could call this function. Don't forget to handle the case when `get_indices` returns `None`.

File: Orange/tests/test_misc.py [hunk 3]
```diff
@@ -28,3 +33,60 @@ def test_memoize_method(self):
...
+
+    def test_create_flagged_table_none_data(self):
+        self.assertIsNone(create_flagged_table(None, None))
+
+    def test_create_flagged_table_none_indices(self):
+        flagged = create_flagged_table(self.zoo, None)
+        self.assertEqual(len(flagged), len(self.zoo))
+        self.assertEqual(0, np.sum([i[FLAGGED_FEATURE_NAME] for i in flagged]))
+
+    def test_cascade_flagged_tables(self):
```
Comment (janezd) [inline Orange/tests/test_misc.py]: I could be too smart by half and replace your function that uses regular expressions with one that uses the first unoccupied name. This test wouldn't fail because there are no "holes". Can you add some code to this test to remove the second meta and then "flag" the table again, so my smart idea would fail?

File: Orange/misc/flagged_data.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import re
+import numpy as np
+from Orange.data import Table, Domain, DiscreteVariable
+
+FLAGGED_SIGNAL_NAME = "Flagged Data"
```
Comment (janezd) [inline Orange/misc/flagged_data.py]: Thinking about it again, I started liking the constant because it will indeed make us stick to the same name in all widgets.

The name is not perfect, though. Not only will we change "flagged" to something else, but it also suggests that it is a name of a flagged signal. `ANNOTATED_DATA_SIGNAL_NAME = "Data"` is better (than `ANNOTATED_SIGNAL_NAME`) but awfully long. Think about it...

Comment (janezd) [inline Orange/misc/flagged_data.py]: If nothing else, the module belongs within `Orange.widgets` because it includes a name of the signal, hence it **is** related to widgets. :)

File: Orange/widgets/unsupervised/tests/test_owdistancemap.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        # check selected data output
+        self.assertIsNone(self.get_output("Data"))
+
+        # check flagged data output
+        flagged = self.get_output(FLAGGED_SIGNAL_NAME)
+        self.assertEqual(0, np.sum([i[FLAGGED_FEATURE_NAME] for i in flagged]))
+
+        # select data points
+        points = random.sample(range(0, len(self.iris)), 20)
```
Comment (janezd) [inline Orange/widgets/unsupervised/tests/test_owdistancemap.py]: I prefer deterministic tests. Randomly choose some indices instead of choosing some indices at random. (https://xkcd.com/221/).

File: Orange/widgets/unsupervised/tests/test_owdistancemap.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        self.widget._selection = points
+        self.widget.commit()
+
+        # check selected data output
+        selected = self.get_output("Data")
+        self.assertEqual(len(selected), len(points))
```
Comment (janezd) [inline Orange/widgets/unsupervised/tests/test_owdistancemap.py]: What about testing that the correct instances were chosen? There may be a better way to do it, but I somewhere used something like `np.testing.assert_almost_equal(selected.X, self.iris.X[points])`.

File: Orange/widgets/unsupervised/tests/test_owmds.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+        # check selected data output
+        self.assertIsNone(self.get_output("Selected Data"))
+
+        # check flagged data output
+        flagged = self.get_output(FLAGGED_SIGNAL_NAME)
+        self.assertEqual(0, np.sum([i[FLAGGED_FEATURE_NAME] for i in flagged]))
+
+        # select data points
+        points = random.sample(range(0, len(self.iris)), 20)
```
Comment (janezd) [inline Orange/widgets/unsupervised/tests/test_owmds.py]: Same as in `DistanceMap`.

**Comment by janezd** [comment]
Github is showing some of my comments as outdated, although you haven't made any further commits. Please check those, too.

Apart from these minor suggestions, I like the PR, in particular your factoring out of the parts of the tests. It would be even greater if you could simulate, say, selection action, but I know this is probably too hard.

Please tell me/us when you make the changes, so we don't wait too long with merging, since rebasing dozens of widgets is not that much fun.

**Comment by VesnaT** [comment]
The comments are outdated because the code was moved to a Mixin in one of the following commits (441b15a), since it was very similar for all widgets. I could have rebased, but wanted to keep the code in case someone didn't like the Mixin idea.
I thought I did simulate the selection...

**Comment by VesnaT** [comment]
Done.

File: Orange/widgets/utils/annotated_data.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    if data is None:
+        return None
+    names = [var.name for var in data.domain.variables + data.domain.metas]
+    name = _get_next_name(names, ANNOTATED_DATA_FEATURE_NAME)
+    metas = data.domain.metas + (DiscreteVariable(name, ("No", "Yes")),)
+    domain = Domain(data.domain.attributes, data.domain.class_vars, metas)
+    annotated = np.zeros((len(data), 1))
+    if selected_indices is not None:
+        annotated[selected_indices] = 1
```
Comment (janezd) [inline Orange/widgets/utils/annotated_data.py]: Should this be 0 or 1? If nothing is selected, all instances have to have Selected=No, no?

File: Orange/widgets/evaluate/tests/test_owconfusionmatrix.py [hunk 9]
```diff
@@ -40,19 +46,11 @@ def test_selected_learner(self):
...
-
-        # check output when data is removed
-        self.send_signal("Evaluation Results", None)
-        self.assertIsNone(self.get_output("Selected Data"))
+        indices = self.widget.tableview.selectedIndexes()
+        indices = {(ind.row() - 2, ind.column() - 2) for ind in indices}
+        selected = [i for i, t in enumerate(zip(
+            self.widget.results.actual, self.widget.results.predicted[0]))
+                    if t in indices]
+        self.selected_indices = self.widget.results.row_indices[selected]
```
Comment (janezd) [inline Orange/widgets/evaluate/tests/test_owconfusionmatrix.py]: Would it be nicer if `_select_data` returned `selected_indices` instead of (ab?)using the instance's attributes for semi-global data storage?

File: Orange/widgets/unsupervised/owhierarchicalclustering.py [hunk 10]
```diff
@@ -1092,10 +1099,12 @@ def commit(self):
 
         if not selected_indices:
             self.send("Selected Data", None)
-            self.send("Other Data", None)
+            annotated_data = create_annotated_table(items, selected_indices) \
```
Comment (janezd) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: Not really an issue, but since you're going to make another commit anyway, can you replace `selected_indices` with `[]`, so that it's obvious this call will always select all (or no) instances.

File: Orange/widgets/unsupervised/owhierarchicalclustering.py [hunk 11]
```diff
@@ -1135,22 +1144,33 @@ def commit(self):
 
             if selected_indices:
                 selected_data = data[mask]
-            if unselected_indices:
-                unselected_data = data[~mask]
+                if self.append_clusters:
+                    def remove_other_value(vars_):
+                        vars_ = [var for var in vars_]
```
Comment (janezd) [inline Orange/widgets/unsupervised/owhierarchicalclustering.py]: Why not `copy`?

File: Orange/widgets/utils/annotated_data.py (hunk 8 above)
Comment (janezd) [inline Orange/widgets/utils/annotated_data.py]: > Should this be 0 or 1? If nothing is selected, all instances have to have Selected=No, no?

I'm stupid. Please ignore.

**Comment by VesnaT** [comment]
I will fix the suggested in the next PR.
